### DISCUSSION TITLE
Multi Tenant App

### DISCUSSION THREAD

**Discussion by jappi00** [filamentphp/filament#2058] [Help] [answered]
Hello together,

I have to accomplish the following task and I want to get some feedback from you.

So I have to write a saas / multi tenant app. I have worked previously a little bit with filament and I love this piece of software so much! So I want to use it.

I have looked into the package https://tenancyforlaravel.com/ and I think it solves many of my problems. But I currently struggel to implement the whole topic.

I have to solve to topics:

1. Admin and Maintain Interface for the saas Topics. (Like: creating new tenants and so on)
2. The App itself. (I also want to use filament for this)

So there are multiple topics and questions in the moment.

1. How can I split the whole filament admin panels in two? What is considered best practice?
(I have already tried to add a middleware for that but had some problems with that.
2. Is this achievable with filament?
3. Is there anything already done in this direction?

Thank you very much!

Greetings,
Jan-Phillip

**Comment by danharrin** [comment]
Why did you want to split the admin in 2? You should probably just hide/show resources and pages using permissions, right?

It is achievable, I've seen people use that package before by adding the middleware to the `config/filament.php` file

**Reply by jappi00** [reply]
Jeah, I think you have a point there. I have the middleware too, but had the problem that I had no access to fillament from the "tenant admin" anymore. I will look into that again. Thank you very much!

**Reply by patrickcurl** [reply]
@jappi00 I'm working on a fork that's to allow multiple dash's ... like one for admin/ one for shop/ and somehow maybe share some nav elements like a menu where you can choose the 'context' or 'view' you want to switch to from a dropdown or something. I also plan to have tenancy built in so you can switch companies/etc. 

My fork is at: https://github.com/zvive/dasher , but it's not usable yet... 

My hope is create multiple dashboards, and integrate some sort of storybook/ui with drag/drop (think webflow-ish), multi-tenancy, saas-features, and merge the frontend bits of something like webflow/crud with backend so us devs can focus more on the backend and less on ui. Also hoping to build a co-op team around using this for client projects, etc...

**Reply by ryangjchandler** [reply]
@patrickcurl If you figure out what needs to be done, please let us know and we'll take a look at the list.

**Reply by ManojKiranA** [reply]
@patrickcurl any update on the package ?

**Reply by thapaPrabhat** [reply]
using `stancl/tenancy` and using `InitializeTenancyByDomain` middleware i can access tenant admin but cannot access admin for landlord to manage tenant or to view aggregate list / reports from tenants.
How can i have filament admin for both landlord app and tenant ?

**Reply by ayann** [reply]
> using `stancl/tenancy` and using `InitializeTenancyByDomain` middleware i can access tenant admin but cannot access admin for landlord to manage tenant or to view aggregate list / reports from tenants.
> How can i have filament admin for both landlord app and tenant ?

Hi @thapaPrabhat did you found a solution? I have the same issue but I don't know how to do.

**Reply by thapaPrabhat** [reply]
Unfortunately no.
Instead i switched to spatie multitenancy package, 
i think if you are using single db its better to write your own logic - which will help to simply things and also configurable to your need.

**Reply by maxroby** [reply]
Same @thapaPrabhat and @ayann issue here, @danharrin no way to get also central app working with stancl/tenancy?

**Reply by ayann** [reply]
@thapaPrabhat @maxroby I found a solution. It wasn't documented, the filament configuration should look like this :

```php
    'middleware' => [
        'auth' => [
            Authenticate::class,
        ],
        'base' => [
            EncryptCookies::class,
            AddQueuedCookiesToResponse::class,
            StartSession::class,
            AuthenticateSession::class,
            ShareErrorsFromSession::class,
            VerifyCsrfToken::class,
            SubstituteBindings::class,
            DispatchServingFilamentEvent::class,
            MirrorConfigToSubpackages::class,
            InitializeTenancyByDomain::class,
            'universal',
        ],
    ],

```

Add the `universal middleware`

**Reply by maxroby** [reply]
Thanks @ayann , very useful! I think it should be documented.

**Reply by jmendozaf** [reply]
With this configuration, I'm able to get the Filament admin panel to work for all tenants as if they were the central domain. Even if the tenant or domain doesn't exist, it always logs in as if it were the central domain. That's not the expected behavior.
The frontend routes for the tenants do work, and I get an error saying the tenant doesn't exist if that's the case.
I think the problem lies in trying to have the Filament admin for both the central domain and the tenants. If I omit the universal middleware in config/filament.php, I get a message saying the tenant cannot be associated with the domain, as already mentioned.

What am I missing here?

**Comment by jmendozaf** [comment]
With this configuration, I'm able to get the Filament admin panel to work for all tenants as if they were the central domain. Even if the tenant or domain doesn't exist, it always logs in as if it were the central domain. That's not the expected behavior.
The frontend routes for the tenants do work, and I get an error saying the tenant doesn't exist if that's the case.
I think the problem lies in trying to have the Filament admin for both the central domain and the tenants. If I omit the universal middleware in config/filament.php, I get a message saying the tenant cannot be associated with the domain, as already mentioned.

What am I missing here?

**Comment by ytorbyk** [comment]
Hi @jappi00 @ayann @maxroby 
Were you able to get working tenancyforlaravel.com with the Filament admin? If you yes, can you please, share what configuration you are using there?
Thanks!

**Reply by luckykenlin** [reply]
Hi @jappi00 @ayann @maxroby @jmendozaf 
I still struggle with this integration. any help would be appreciate

**Reply by Arpityadav** [reply]
Hi! Did you figure it out? Appreciate the help.

**Comment by luckykenlin** [comment]
Hi @jappi00 @ayann @maxroby @jmendozaf 
I still struggle with this integration. any help would be appreciate

**Reply by Arpityadav** [reply]
Hey! If you find a way to make it work, let me know

**Comment by coolsam726** [comment]
I have successfully setup Filament to work with [tenancyforlaravel.com](https://tenancyforlaravel.com/), but I am using Filament 3.x and Livewire 3.x. I have a central admin panel, a tenant's admin panel and a tenant's portal panel as separate filament panels. Let me know if you are interested in the configuration.

**Reply by jappi00** [reply]
Are you using separate databases?

**Reply by ytorbyk** [reply]
@coolsam726 Awesome! Yes, very interested. Can you share your configuration? 
Thanks!

**Reply by jmendozaf** [reply]
Great. Can you share some workaround? Are you using different databases or schemas?

**Reply by jappi00** [reply]
FYI: maybe this helps: https://gist.github.com/nathandaly/44a83df5e3ea735eae91652c7a4894c9

**Reply by luckykenlin** [reply]
@coolsam726 Are you using separate databases? Can you share your configuration? thanks

**Reply by coolsam726** [reply, marked as answer]
Sorry for the delay.
Yes, I am using PostgreSQL so I preferred to use the option of using a single Database but different schemas, one schema per tenant. I am also using identification by subdomain. 
For ease of reference, I have decided to share my repo publicly as a starter code for anyone interested. Here is the link.
https://github.com/savannabits/filament-tenancy-starter

Pay attention to the following:

1. I have three Filament Panel Providers:
* AdminPanelProvider is the Admin panel for the central admin
* TenantAdminPanelProvider is the admin panel for each tenant
* TenantPortalPanelProvider is the customer facing portal for each tenant.

2. For **AdminPanelProvider**, I am using the `universal` middleware as part of the middleware. Most importantly, I am also setting the Panel's domains as any of the Central Domains. This will identify the panel as a Central Panel.

3. For the Tenant Panels, I am using the `universal`,  `InitializeTenancyBySubdomain` and `PreventAccessFromCentralDomains` middleware. 
4. Check the **TenancyServiceProvider** and see that under `prepareLivewireForTenancy()` method I am setting a custom Livewire **updateRoute** which includes the `universal` and `InitializeTenancyBySubdomain` middleware in addition to the `web` middleware.

I don't believe I have left out any important info, but if you have any questions do let me know. Happy Coding!

**Reply by jappi00** [reply]
What do you mean by different schemas?

**Reply by jmendozaf** [reply]
One database, different pg schema for tenant

**Reply by jappi00** [reply]
Ah I wasn't aware what the schema system in Postgres really is.

**Reply by coolsam726** [reply]
I find it neater as opposed to spinning up an entire DB for each tenant. Either way, this will still work even if you were to use different pgsql or mysql databases for each tenant.

**Reply by madushancs** [reply]
Hi coolsam726, Is this working with mysql?

**Reply by coolsam726** [reply]
I haven't tested yet but I don't see why not. You just need to change the DB_ env  variables to match your connection. MYSQL doesn't support multiple schemas per database, so in that case the app will spin up one database for each tenant.

**Reply by iamtomc** [reply]
Well done @coolsam726 this what I have been looking for a while now thanks a ton.

**Reply by citricguy** [reply]
@coolsam726 you're amazing, thank you. This is exactly what I was looking for.

**Reply by rasul-chopurov** [reply]
419 error in portal or admin

i am use you starter

**Reply by citricguy** [reply]
If I recall, I had to change my SESSION_DRIVER to something other than database. I ended up on Redis and the 419's seemed to clear up. I also had to set 'asset_helper_tenancy' to false inside config/tenancy.php in that same commit. Not sure if related.

**Reply by Waguilar33** [reply]
Getting the following error when creating a new Tenant

`Class "Redis" not found`

**Reply by ryangjchandler** [reply]
@Waguilar33 Make sure you've configured your `QUEUE_CONNECTION` correctly.

**Reply by sudhansu08** [reply]
@coolsam726 what is the Core plugin that you are using. what does this do and do we need this?
